#FIAP
#PÓS-GRADUAÇÃO – IA PARA DEVS
#Pós Tech – 1IADT
#TECH CHALLENGE – FASE 2

Grupo 2:
* Gerdson Cunha Pinheiro
* Rafael Valentim Fonseca
* Rogério Maia de Queiroz Lessa
* Vanderci José Colasso
* Wesley Aoyama Silva


# Descrição do Problema
A roteirização da entrega de cartas/encomendas dos Correios é uma situação que acontece diariamente nas atividades operacionais da empresa. A otimização do percurso é essencial para o cumprimento das metas da empresa em relação à qualidade do serviço prestado ao cliente.

Atualmente a Empresa de Correios e Telégrafos conta com mais de 30.000 carteiros realizando entregas de encomendas em todo o Brasil.  As entregas são triadas por distritos postais que, por sua vez, possuem uma rota específica de entrega. As rotas seguem uma ordem definida previamente pelos gestores das unidades de entrega.  Antes de sair para realizar as entregas dos objetos, o carteiro organiza os objetos de acordo com a ordem de entrega para facilitar o trabalho na rua. Durante as entregas, o sistema App Operacional, responsável pelas entregas, manda informações georreferenciadas para um banco de dados corporativo para uso posterior.

O problema enfrentado pelas unidades é a mesma do caixeiro viajante que necessita escolher o menor caminho para percorrer em uma determinada região. A empresa conta com milhares de distritos com muitos pontos de entrega que dependem do conhecimento empírico dos gestores na escolha do melhor percurso a ser percorrido pelos entregadores.

Implementação do algoritmo

Com os ensinamentos aprendidos nesta fase 2, detectamos que uma solução possível para tentar resolver o problema seria o uso do algoritmo genético para otimizar essas roteirizações, com base nas informações dos distritos e dos pontos de entrega. Essa seria uma solução inspirada no problema do cacheiro viajante, citado nas aulas.

Simulamos um roteiro com algumas localizações aleatórias dentro de um mesmo município, sempre partindo de um Centro de Distribuição Domiciliar (CDD) dos Correios.  Calculando as distâncias usando uma API do Google Maps, pelas coordenadas geodésicas de cada ponto, conseguimos inserir na biblioteca DEAP, que fazendo os cálculos dentro das gerações, conseguimos chegar a um roteiro otimizado para os carteiros efetuarem seu serviço.

# Otimização de Rotas de Entrega com Algoritmos Genéticos
Este código implementa um algoritmo genético para otimizar a rota de entrega de um carteiro, utilizando a biblioteca DEAP e a API do Google Maps para calcular distâncias. O problema é análogo ao problema do Caixeiro Viajante, onde o objetivo é encontrar o caminho mais curto para visitar todos os pontos de entrega e retornar à origem.
1. Imports e Inicialização

- Importa as bibliotecas necessárias: random, numpy, deap e googlemaps.
- Define a chave da API do Google Maps para calcular distâncias.
- Cria um objeto gmaps para interagir com a API.

In [ ]:
!pip install deap

!pip install googlemaps

import random
import numpy as np
from deap import algorithms, base, creator, tools
import googlemaps

# Chave da API do Google Maps
API_KEY = "AIzaSyD267_5UL_l5KyXmuCmEtY395cvj7hpQbA"
gmaps = googlemaps.Client(key=API_KEY)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.4/135.4 kB 2.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for googlemaps: filename=googlemaps-4.10.0-py3-none-any.whl size=40712 sha256=7a392cbffa9812cc3d98a8edcf442f8d24dfbd46839671c50b7ae1a8064cec63
  Stored in directory: /root/.cache/pip/wheels/17/f8/79/999d5d37118fd35d7219ef57933eb9d09886c4c4503a800f84
Successfully built googlemaps


2. Classe Localizacao

- Define uma classe Localizacao para armazenar informações sobre cada ponto de entrega: nome, coordenadas (latitude e longitude) e CEP.

In [ ]:
class Localizacao:
    """
    Representa uma localização com nome, coordenadas e CEP.

    Attributes:
        nome (str): Nome da localização.
        coordenadas (tuple): Tupla com latitude e longitude.
        cep (str): CEP da localização.
    """
    def __init__(self, nome, coordenadas, cep):
        self.nome = nome
        self.coordenadas = coordenadas
        self.cep = cep

3. Função calcular_distancia

- Define uma função calcular_distancia para calcular a distância entre duas coordenadas usando a API do Google Maps.
- Utiliza o método distance_matrix da API para obter a distância em metros.
- Retorna float("inf") em caso de erro.

In [ ]:
def calcular_distancia(origem, destino):
    """
    Calcula a distância entre duas coordenadas usando a API do Google Maps.

    Args:
        origem (str ou tuple): Coordenadas de origem.
        destino (str ou tuple): Coordenadas de destino.

    Returns:
        int: Distância em metros entre a origem e o destino.
        float("inf"): Retorna infinito em caso de erro.
    """
    try:
        # Tenta calcular a distância usando a API do Google Maps
        matrix = gmaps.distance_matrix(origins=origem, destinations=destino, mode="driving")

        # Extrai a distância em metros da resposta da API
        distancia = matrix["rows"][0]["elements"][0]["distance"]["value"]
        return distancia

    except Exception as e:
        print(f"Erro ao calcular a distância: {e}")
        return float("inf")  # Retorna infinito em caso de erro

4. Definição de Localizações

- Define a localização de origem (origem) como o Centro de Distribuição Domiciliar (CDD).
- Define uma lista localizacoes com as localizações de entrega.

In [ ]:
# Localizacoes

# Define a origem (Centro de Distribuição Domiciliar - CDD)
origem = Localizacao("CDD Bequimao", (-2.5373686770484736, -44.22756163011021), "65060-971")

# Define as localizações de entrega
localizacoes = [
    Localizacao("Forquilha", (-2.554292, -44.281667), "65085-630"),
    Localizacao("Jardim São Cristóvão", (-2.544834, -44.308271), "65076-520"),
    Localizacao("Cohatrac", (-2.539701942265006, -44.20967851718005), "65054-270"),
    Localizacao("Parque Vitória", (-2.519003449355086, -44.208967029288345), "65073-525"),
    Localizacao("Cohab", (-2.5455106589603727, -44.21923062005139), "65073-525")
]

5. Função gerar_matriz_distancia

- Define uma função gerar_matriz_distancia para criar uma matriz de distâncias entre todos os pontos de entrega, incluindo a origem.
- Calcula as distâncias usando a função calcular_distancia.
- Retorna a matriz de distâncias.

In [ ]:
def gerar_matriz_distancia(origem, localizacoes):
  """
  Gera uma matriz de distâncias entre a origem e as localizações.

  Args:
      origem (Localizacao): Objeto Localizacao representando a origem.
      localizacoes (list): Lista de objetos Localizacao representando os destinos.

  Returns:
      numpy.ndarray: Matriz de distâncias.
  """

  n = len(localizacoes) + 1
  dist = np.zeros((n, n), dtype=int)

  for i in range(1, n):
    dist[0][i] = calcular_distancia(origem.coordenadas, localizacoes[i-1].coordenadas)
    dist[i][0] = calcular_distancia(localizacoes[i-1].coordenadas, origem.coordenadas)
    for j in range(1, n):
      if i != j:
        dist[i][j] = calcular_distancia(localizacoes[i-1].coordenadas, localizacoes[j-1].coordenadas)

  return dist

6. Função avaliar_individuo

- Define uma função avaliar_individuo para calcular a distância total de um caminho circular, dado um indivíduo (sequência de índices de localizações) e a matriz de distâncias.
- Percorre a sequência de índices e soma as distâncias entre os pontos correspondentes na matriz.
- Retorna a distância total como uma tupla.

In [ ]:
def avaliar_individuo(individuo, dist):
  """
  Avalia a qualidade de um indivíduo (sequência de localizações) calculando a distância total do caminho circular.

  Args:
      individuo (list): Sequência de índices de localizações, representando um caminho.
      dist (numpy.ndarray): Matriz de distâncias.

  Returns:
      tuple: Tupla contendo a distância total.
  """

  n = len(individuo)
  distancia_total = dist[0][individuo[0]+1]
  for i in range(n - 1):
    local1 = individuo[i] + 1
    local2 = individuo[i+1] + 1
    distancia = dist[local1][local2]
    distancia_total += distancia

  distancia_total += dist[individuo[-1]+1][0]

  return distancia_total,

7. Configuração do Algoritmo Genético

- creator.create("FitnessMin", base.Fitness, weights=(-1.0,)): Define um novo tipo de dado chamado FitnessMin que herda da classe base.Fitness e define um peso de -1.0 para a aptidão. Isso significa que o algoritmo genético buscará minimizar o valor da aptidão, que neste caso representa a distância total da rota.

In [ ]:
# Define o tipo de dado para a aptidão (minimizar a distância)
creator.create("FitnessMin", base.Fitness, weights=(-1.0,))

- creator.create("Individuo", list, fitness=creator.FitnessMin): Define um novo tipo de dado chamado Individuo que é uma lista (representando a ordem das localizações) e possui um atributo de aptidão do tipo FitnessMin.

In [ ]:
# Define o tipo de dado para um indivíduo (sequência de localizações)
creator.create("Individuo", list, fitness=creator.FitnessMin)

- tamanho_populacao = 10: Define o tamanho da população inicial do algoritmo genético como 10 indivíduos.
- probabilidade_crossover = 0.8: Define a probabilidade de crossover entre indivíduos como 0.8, ou seja, 80% dos indivíduos serão cruzados em cada geração.
- probabilidade_mutacao = 0.3: Define a probabilidade de mutação de um indivíduo como 0.3, ou seja, 30% dos indivíduos sofrerão mutação em cada geração.
- numero_geracoes = 100: Define o número de gerações do algoritmo genético como 100.

In [ ]:
# Define os parâmetros do algoritmo genético
tamanho_populacao = 10
probabilidade_crossover = 0.8
probabilidade_mutacao = 0.2
numero_geracoes = 100

- toolbox = base.Toolbox(): Cria um objeto toolbox do tipo base.Toolbox, que é um gerenciador de funções do DEAP. Esse objeto armazenará as funções necessárias para o algoritmo genético, como as funções de criação de indivíduos, população, avaliação, cruzamento, mutação e seleção.

In [ ]:
# Cria o ambiente DEAP
toolbox = base.Toolbox()

- toolbox.register("indices", random.sample, range(len(localizacoes)), len(localizacoes)): Registra uma função chamada indices no toolbox. Essa função é definida como random.sample, que recebe dois argumentos: range(len(localizacoes)), que gera uma lista de números de 0 a len(localizacoes)-1 (o número de localizações), e len(localizacoes), que define o número de elementos que serão amostrados aleatoriamente da lista gerada. Essa função irá gerar uma lista aleatória de índices que representam a ordem de visita das localizações.

- toolbox.register("individuo", tools.initIterate, creator.Individuo, toolbox.indices): Registra uma função chamada individuo no toolbox. Essa função é definida como tools.initIterate, que recebe três argumentos: creator.Individuo, que define o tipo de dado do indivíduo, toolbox.indices, que é a função que gera a lista de índices, e creator.Individuo, que define o tipo de dado do indivíduo. Essa função irá criar um novo indivíduo a partir da lista de índices gerada pela função indices.

In [ ]:
# Define a função de criação de indivíduos
toolbox.register("indices", random.sample, range(len(localizacoes)), len(localizacoes))
toolbox.register("individuo", tools.initIterate, creator.Individuo, toolbox.indices)

- toolbox.register("populacao", tools.initRepeat, list, toolbox.individuo): Registra uma função chamada populacao no toolbox. Essa função é definida como - tools.initRepeat, que recebe três argumentos: list, que define o tipo de dado da população (uma lista de indivíduos), toolbox.individuo, que é a função que cria um novo indivíduo, e tamanho_populacao, que define o número de indivíduos na população. Essa função irá criar a população inicial do algoritmo, utilizando a função individuo para gerar cada indivíduo da população.

In [ ]:
# Define a função de criação da população
toolbox.register("populacao", tools.initRepeat, list, toolbox.individuo)

- dist = gerar_matriz_distancia(origem, localizacoes): Chama a função gerar_matriz_distancia para gerar a matriz de distâncias entre todos os pontos de entrega, incluindo a origem. A matriz é armazenada na variável dist.

In [ ]:
# Gera a matriz de distancias
dist = gerar_matriz_distancia(origem, localizacoes)
print(dist)

[[    0 10305 21588  2958  3187  2053]
 [ 8174     0 14721 11131 11360  9264]
 [19684 13615     0 22641 22871 20774]
 [ 4487 12327 23306     0  4317  1610]
 [ 4509 13451 23669  4412     0  4592]
 [ 2584 10445 21727  2048  4957     0]]


- toolbox.register("evaluate", avaliar_individuo, dist=dist): Registra uma função chamada evaluate no toolbox. Essa função é definida como avaliar_individuo, que recebe dois argumentos: individuo, que é o indivíduo a ser avaliado, e dist, que é a matriz de distâncias. Essa função irá calcular a distância total da rota representada pelo indivíduo e retornar o valor como a aptidão do indivíduo.
- toolbox.register("mate", tools.cxOrdered): Registra uma função chamada mate no toolbox. Essa função é definida como tools.cxOrdered, que é um operador de crossover para indivíduos que são representados por listas ordenadas. Esse operador garante que os filhos gerados pelo crossover também sejam listas ordenadas, respeitando a ordem das localizações.
- toolbox.register("mutate", tools.mutShuffleIndexes, indpb=0.05): Registra uma função chamada mutate no toolbox. Essa função é definida como tools.mutShuffleIndexes, que é um operador de mutação para indivíduos que são representados por listas ordenadas. Esse operador embaralha os índices da lista do indivíduo com uma probabilidade definida por indpb, que é 0.05 neste caso.
- toolbox.register("select", tools.selTournament, tournsize=3): Registra uma função chamada select no toolbox. Essa função é definida como tools.selTournament, que é um método de seleção de indivíduos para a próxima geração, baseado em um torneio. O argumento tournsize define o tamanho do torneio, que é 3 neste caso, ou seja, 3 indivíduos são selecionados aleatoriamente para competir em cada torneio. O indivíduo com a melhor aptidão é então selecionado para a próxima geração.

In [ ]:
# Define as funções de avaliação, cruzamento, mutação e seleção
toolbox.register("evaluate", avaliar_individuo, dist=dist)
toolbox.register("mate", tools.cxOrdered)
toolbox.register("mutate", tools.mutShuffleIndexes, indpb=0.05)
toolbox.register("select", tools.selTournament, tournsize=3)


- populacao = toolbox.populacao(n=tamanho_populacao): Cria a população inicial do algoritmo genético, utilizando a função populacao registrada no toolbox. A população é composta por tamanho_populacao indivíduos, cada um criado utilizando a função individuo.

In [ ]:
# Cria a população inicial
populacao = toolbox.populacao(n=tamanho_populacao)

8. Estatísticas e Hall of Fame

- Define uma função salvarEstatistica para retornar a aptidão de um indivíduo.
- Configura as estatísticas para o algoritmo genético, incluindo a estatística min para a aptidão mínima.
- Define o hof (Hall of Fame) para guardar o melhor indivíduo encontrado.

In [ ]:
# Funcao para salvar as aptidoes de cada individuo
def salvarEstatistica(individuo):
  return individuo.fitness.values

# https://deap.readthedocs.io/en/master/api/tools.html#deap.tools.Statistics
# Configura as estatísticas para o algoritmo genético

estatistica = tools.Statistics(salvarEstatistica)
estatistica.register('min', np.min)

# https://deap.readthedocs.io/en/master/api/tools.html#hall-of-fame
# Define o Hall of Fame para guardar o melhor indivíduo - Elitismo

hof = tools.HallOfFame(1)

9. Execução do Algoritmo Genético

- Executa o algoritmo genético eaMuPlusLambda com os parâmetros definidos.
- Essa função implementa a estratégia de seleção $(\mu + \lambda)$, que combina elitismo com a geração de novos indivíduos. Essa estratégia gera lambda_ novos indivíduos (filhos) a partir da população atual (mu) através de crossover e mutação. Então, a população da próxima geração é formada selecionando os mu melhores indivíduos da população atual e os lambda_ filhos. Essa técnica garante que os melhores indivíduos da população anterior sejam preservados na próxima geração.
- populacao: A população inicial do algoritmo genético.
- toolbox: O objeto toolbox contendo as funções registradas para o algoritmo (criação de indivíduos, avaliação, crossover, mutação, etc.).
- mu: O tamanho da população atual (mu = tamanho_populacao).
- lambda_: O número de novos indivíduos gerados em cada geração (lambda_ = tamanho_populacao).
- cxpb: A probabilidade de crossover entre indivíduos (cxpb = probabilidade_crossover).
- mutpb: A probabilidade de mutação de um indivíduo (mutpb = probabilidade_mutacao).
- stats: O objeto estatistica que coleta e registra estatísticas sobre o algoritmo (como a aptidão mínima).
- ngen: O número de gerações para executar o algoritmo (ngen = numero_geracoes).
- halloffame: O objeto hof (Hall of Fame) que armazena o melhor indivíduo encontrado até o momento.
- verbose=True: Define que o algoritmo deve imprimir informações sobre o progresso durante a execução.
- Salva o resultado e o log do algoritmo.

In [ ]:
# Executa o algoritmo genético
resultado, log = algorithms.eaMuPlusLambda(
    populacao,
    toolbox,
    mu=tamanho_populacao,
    lambda_=tamanho_populacao,
    cxpb=probabilidade_crossover,
    mutpb=probabilidade_mutacao,
    stats=estatistica,
    ngen=numero_geracoes,
    halloffame=hof,
    verbose=True,
)

gen	nevals	min  
0  	10    	55383
1  	10    	55383
2  	10    	55383
3  	10    	55383
4  	10    	53876
5  	10    	53876
6  	10    	53876
7  	10    	53876
8  	10    	53876
9  	10    	53876
10 	10    	53876
11 	10    	53876
12 	10    	53876
13 	10    	53876
14 	10    	53876
15 	10    	53876
16 	10    	53876
17 	10    	53876
18 	10    	53876
19 	10    	53876
20 	10    	53876
21 	10    	53876
22 	10    	53876
23 	10    	53876
24 	10    	53876
25 	10    	53876
26 	10    	53876
27 	10    	53876
28 	10    	53876
29 	10    	53876
30 	10    	53876
31 	10    	53876
32 	10    	53876
33 	10    	53876
34 	10    	53876
35 	10    	53876
36 	10    	52725
37 	10    	52725
38 	10    	52725
39 	10    	52725
40 	10    	52725
41 	10    	52725
42 	10    	52725
43 	10    	52725
44 	10    	52725
45 	10    	52725
46 	10    	52725
47 	10    	52725
48 	10    	52725
49 	10    	52725
50 	10    	52725
51 	10    	52725
52 	10    	52725
53 	10    	52725
54 	10    	52725
55 	10    	52725
56 	10    	52725
57 	10    	527

10. Impressão dos Resultados

- Imprime a última população (indivíduos da última geração).
- Imprime o melhor indivíduo encontrado (Hall of Fame).
- Imprime a aptidão (distância total) do melhor indivíduo.
- Imprime o log do algoritmo.
- Obtém o melhor indivíduo da última população.
- Imprime o melhor caminho encontrado, incluindo a origem e o destino.
- Imprime a distância total do melhor caminho.

In [ ]:
# Ultima populacao (individuos da ultima geracao)
print(resultado)

[[3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0], [3, 2, 4, 1, 0]]


In [ ]:
# Melhor rota/individuo
print(hof)

[[3, 2, 4, 1, 0]]


In [ ]:
# Aptidao da melhor rota
melhor = hof[0]
print(avaliar_individuo(melhor, dist))

(52725,)


In [ ]:
# Log
log

[{'gen': 0, 'nevals': 10, 'min': 55383.0},
 {'gen': 1, 'nevals': 10, 'min': 55383.0},
 {'gen': 2, 'nevals': 10, 'min': 55383.0},
 {'gen': 3, 'nevals': 10, 'min': 55383.0},
 {'gen': 4, 'nevals': 10, 'min': 53876.0},
 {'gen': 5, 'nevals': 10, 'min': 53876.0},
 {'gen': 6, 'nevals': 10, 'min': 53876.0},
 {'gen': 7, 'nevals': 10, 'min': 53876.0},
 {'gen': 8, 'nevals': 10, 'min': 53876.0},
 {'gen': 9, 'nevals': 10, 'min': 53876.0},
 {'gen': 10, 'nevals': 10, 'min': 53876.0},
 {'gen': 11, 'nevals': 10, 'min': 53876.0},
 {'gen': 12, 'nevals': 10, 'min': 53876.0},
 {'gen': 13, 'nevals': 10, 'min': 53876.0},
 {'gen': 14, 'nevals': 10, 'min': 53876.0},
 {'gen': 15, 'nevals': 10, 'min': 53876.0},
 {'gen': 16, 'nevals': 10, 'min': 53876.0},
 {'gen': 17, 'nevals': 10, 'min': 53876.0},
 {'gen': 18, 'nevals': 10, 'min': 53876.0},
 {'gen': 19, 'nevals': 10, 'min': 53876.0},
 {'gen': 20, 'nevals': 10, 'min': 53876.0},
 {'gen': 21, 'nevals': 10, 'min': 53876.0},
 {'gen': 22, 'nevals': 10, 'min': 53876.0}

In [ ]:
# Obtém o melhor indivíduo (menor distância)
melhor_individuo = tools.selBest(resultado, k=1)[0]
melhor_individuo

[3, 2, 4, 1, 0]

In [ ]:
# Imprime o melhor caminho e a distância total

print("Melhor caminho encontrado:")
print(origem.nome)
for i in melhor_individuo:
  print(localizacoes[i].nome)

print(origem.nome)

Melhor caminho encontrado:
CDD Bequimao
Parque Vitória
Cohatrac
Cohab
Jardim São Cristóvão
Forquilha


In [ ]:
distancia_total = avaliar_individuo(melhor_individuo, dist)
print(f"Distância total: {distancia_total[0]} metros")

CDD Bequimao
Distância total: 52725 metros


# Gerando arquivo geojson para visualizar a rota

Ao término da geração das informações referentes ao algoritmo genético, foi utilizado o padrão GeoJson para gerar as informações georreferenciadas do resultado obtido.

GeoJSON é uma iniciativa open-source, criada para estruturar em forma de código as estruturas geográficas que compõem o planeta terra. Baseado no JSON (JavaScript Object Notation), comumente utilizado em APIs REST. O GeoJson utiliza o formato Json como padrão.

Segue um link da Medium explicando sobre esse padrão: https://medium.com/academy-eldoradocps/o-que-%C3%A9-e-como-funciona-o-geojson-43aca509fe4d

Trata-se de um padrão internacional adotado comumente no mercado para disponibilizar informações geográficas em diversos países.

O Python, por sua vez, possui bibliotecas que permite a geração desse padrão.

In [ ]:
! pip install geojson #Adicionando a dependência para gerar o geojson.
import geojson

In [ ]:
#Criando uma lista de coordenadas e outra de features (padrão do geojson)
coord = []
features =[]

 No problema do caixeiro viajante, o percurso sempre começa e termina na origem.

A lista coord[ ] é responsável por ligar, por meio de linhas retas, os pontos de entrega.

Sendo assim, no array de coordenadas (coord) o primeiro e o última elemento deve ser a coordenada da origem.


In [ ]:
coord.append(origem.coordenadas[::-1])


features.append(geojson.Feature(
        geometry=geojson.Point(origem.coordenadas[::-1]), #Invertendo os valores das coordenadas.
        properties= list({"name", origem.nome})
    ))

for i in melhor_individuo:
    print(localizacoes[i].nome)

    coord.append(list(localizacoes[i].coordenadas[::-1]))

    features.append(geojson.Feature(
        geometry=geojson.Point(localizacoes[i].coordenadas[::-1]), #Invertendo os valores das coordenadas.
        properties= list({"name", localizacoes[i].nome})
    ))

coord.append(origem.coordenadas[::-1])

features.append(geojson.Feature(
     geometry=geojson.LineString(list(coord)),
     properties=list({"name", "rota"})
))

Parque Vitória
Cohatrac
Cohab
Jardim São Cristóvão
Forquilha


No código acima, o objeto features recebe dois tipos de parâmetro:


*   O parâmetro Point que é responsável por plotar os pontos de entrega no mapa;
*   O parâmetro LineString que liga os pontos de entrega formando a rota;

Foi utilizada a mesma iteração com o melhor indivíduo para montar as informações do GeoJson.

Após o parâmetro LineString é gerado ao final da iteração com todas as coordenadas.

In [ ]:
# Cria uma FeatureCollection
feature_collection = geojson.FeatureCollection(list(features))

# Escreve o GeoJSON em um arquivo
with open('example.geojson', 'w') as f:
    geojson.dump(feature_collection, f)

De posse do arquivo gerado, com o nome example.geojson, este sejá importado na plataforma https://geojson.io/ para que seja possível a visualização da rota definida pelo algoritmo genético conforme link disponibilizado abaixo:

https://app.screencastify.com/v3/watch/Vd7fOMIJjAMQrU7Ao71l


# Conclusões
Os resultados mostram que o algoritmo genético convergiu para um valor mínimo de 52725 após algumas iterações. É possível observar que:
- A distância mínima encontrada foi 52725 metros. Este valor ficou constante a partir da geração 36, indicando que o algoritmo encontrou um ótimo local(pode não ser a melhor solução).
- O número de avaliações (nevals) por geração é constante em 10. Isso significa que o algoritmo está avaliando 10 indivíduos por geração.
- O algoritmo demorou algumas gerações para encontrar uma solução significativamente melhor. A distância mínima ficou em torno de 54922 metros por 36 gerações, e então, de repente, caiu para 52725 metros.

### Considerações:
- Algoritmos híbridos: Podemos misturar diversas estratégias para determinados cenários, como por exemplo, para cenários com poucas combinações, podemos utilizar um algoritmo de força bruta(garante a melhor solução sempre).
- Parâmetros do algoritmo: Os parâmetros do algoritmo (tamanho da população, probabilidade de crossover e mutação, número de gerações) podem influenciar o desempenho do algoritmo. Neste caso, uma população maior poderia ter levado a uma convergência mais rápida.
- Complexidade do problema: A complexidade do problema de roteirização pode impactar a capacidade do algoritmo de encontrar uma solução ótima. Se o número de pontos de entrega for muito grande, o algoritmo pode levar mais tempo para convergir ou pode não encontrar a melhor solução.
- Precisão da API do Google Maps: A precisão das distâncias calculadas pela API do Google Maps pode afetar a precisão da solução encontrada. É importante considerar que a API pode retornar valores aproximados, e isso pode influenciar a qualidade da solução final.

### Possíveis ações para utilização em ambiente de produção:
- Aumentar o tamanho da população: Uma população maior pode aumentar a diversidade genética e melhorar as chances de encontrar uma solução melhor.
- Ajustar a probabilidade de crossover e mutação: Ajuste fino desses parâmetros pode levar a uma exploração mais eficiente do espaço de busca.
- Aumentar o número de gerações: Mais gerações permitem que o algoritmo explore mais soluções, aumentando a chance de encontrar um ótimo global.
- Usar um algoritmo genético mais sofisticado: Existem outros algoritmos genéticos mais complexos que podem oferecer melhores resultados.
- Usar dados mais precisos: Se possível, utilizar dados de localização mais precisos (como coordenadas GPS) pode levar a resultados mais precisos.

Os resultados obtidos indicam que o algoritmo genético conseguiu encontrar uma solução razoável para o problema de roteirização, mas é possível que existam soluções melhores. Ajustar os parâmetros e explorar outros algoritmos podem levar a resultados mais precisos e eficientes. A "queda" repentina na distância mínima na geração 65 sugere que o algoritmo pode ter encontrado um ótimo local, mas é importante realizar mais testes para confirmar essa hipótese.

Este código demonstra como utilizar algoritmos genéticos para otimizar a rota de entrega de um carteiro. O algoritmo encontra uma rota eficiente (nem sempre a melhor), minimizando a distância total percorrida. Este é apenas um cenário básico, o código pode ser adaptado para lidar com cenários mais complexos nos sistemas dos Correios, incluindo diferentes tipos de veículos, restrições de tempo, janelas de entrega e outros fatores que podem influenciar a rota ideal.